src/Microsoft.OpenApi/Services/OpenApiUrlTreeNode.cs

```diff
@@ -235,5 +236,152 @@ public void AddAdditionalData(Dictionary<string, List<string>> additionalData)
+                var (childStartChar, childEndChar) = GetShapeDelimiters(childMethods);
+                writer.WriteLine($"{path}{startChar}\"{node.Segment}\"{endChar} --> {SanitizeMermaidNode(child.Value.Path)}{childStartChar}\"{child.Key}\"{childEndChar}");
+                ProcessNode(child.Value, writer);
+            }
+            if (String.IsNullOrEmpty(methods)) methods = "OTHER";
+            writer.WriteLine($"class {path} {methods}");
+        }
+
+        private static string GetMethods(OpenApiUrlTreeNode node)
+        {
+            return String.Join("_", node.PathItems.SelectMany(p => p.Value.Operations.Select(o => o.Key))
+                .Distinct()
+                .Select(o => o.ToString().ToUpper())
+                .OrderBy(o => o)
+                .ToList());
+        }
+
+        private static (string, string) GetShapeDelimiters(string methods)
+        {
+            
+            if (MermaidNodeStyles.ContainsKey(methods))
+            {
+                //switch on shape
+                switch (MermaidNodeStyles[methods].Shape)
+                {
+                    case MermaidNodeShape.Circle:
+                        return ("((", "))");
+                    case MermaidNodeShape.RoundedCornerRectangle:
+                        return ("(", ")");
+                    case MermaidNodeShape.Rhombus:
+                        return ("{", "}");
+                    case MermaidNodeShape.SquareCornerRectangle:
+                        return ("[", "]");
+                    case MermaidNodeShape.OddShape:
+                        return (">", "]");
+                    default:
+                        return ("[", "]");
+                }
+            }
+            else
+            {
+                return ("[", "]");
+            }
+        }
+        private static string SanitizeMermaidNode(string token)
+        {
+            return token.Replace("\\", "/")
+                    .Replace("{", ":")
+                    .Replace("}", "")
+                    .Replace(".", "_")
+                    .Replace(";", "_")
+                    .Replace("-", "_")
+                    .Replace("default", "def_ault");  // default is a reserved word for classes
+        }
+    }
+    /// <summary>
+    /// Defines the color and shape of a node in a Mermaid graph diagram
+    /// </summary>
+    public class MermaidNodeStyle
+    {
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="color"></param>
+        /// <param name="shape"></param>
+        public MermaidNodeStyle(string color, MermaidNodeShape shape)
+        {
+            Color = color;
+            Shape = shape;
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public string Color { get;  }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public MermaidNodeShape Shape { get;  }
+    }
+
+    /// <summary>
+    /// 
+    /// </summary>
+    public enum MermaidNodeShape
```

Comment:
does it need to be public?

Reply:
Unfortunately yes, even though I have made it read-only.  Hidi needs to read the list of styles to be able to generate the HTML legend for the diagrams.